### ISSUE TITLE
EditorPlugin: no way to know when the current object is deselected

### ISSUE THREAD

**Issue by Zylann** [godotengine/godot#6459]
Godot 2.1:

I see there is an edit() callback in EditorPlugin, however it is only called when the handled object is selected. When something else gets selected there is no event. I thought edit() would be called with null, but it isn't.

So far I also relied on the Selection object, and it worked pretty well for nodes, so I can hide my editor panel depending on that.

But now I realize it's not great, because you can edit something that is not a node, a shader for example: https://github.com/Zylann/godot_terrain_plugin/issues/8
![Screenshot](https://camo.githubusercontent.com/6a6854ad5edd157dcfb4830d52bd1510548a755e/68747470733a2f2f692e696d6775722e636f6d2f546a6a3978434a2e706e67)

How do I tell my panel to hide in this case?
Doing that in handles() is not appropriate, as it's a testing function, it's not meant to do any action.
I would expect edit() to be called with a null argument, or have a stop_edit() callback.
If all of this is intented, then maybe the panels should be stacked, but I don't think it makes sense.
Finally, if there is already a way to do that, please let me know.

**Comment by GlaDOSik** [comment]
Take a look at documentation again ;) I'm sure `make_visible(visible)` callback does exactly what you need. About resource editing, my plugin edits Mesh resource so I check if edited MeshInstance has this resource. If you edit the shader, it still should be assigned to some surface in Mesh which is in node you get from EditorSelection.

**Comment by Zylann** [comment]
Thanks I will try with make_visible() :)

I don't understand why my pluging should care about where is the edited shader. Drilling up/down from the node to check that sounds not convenient.

**Comment by GlaDOSik** [comment]
Yeah, it would be convenient to have selected resource in selection too. When make_visible argument is false, it means you won't handle object so it doesn't matter what your reference is.

**Comment by Zylann** [comment]
Yeah, but I'm just wondering that just in case I make a Reference editor (resource?), because it could produce a leak if I don't release the reference.

**Comment by GlaDOSik** [comment]
That could be a problem only if you store your references in array and never releasing them. If you do something like this, I wouldn't be too worried about leak.
```
var referenceToResource
func edit(object):
referenceToResource = object
```

**Comment by Zylann** [comment]
My plugin is probably going to be able to edit heightfields, which are resources as big as a 1024x1024 32-bit image.
I am already doing this with `edit`, however because it never gets called when the resource is deselected, I have no chance to know when to set my reference to null, which is a bit problematic for such large objects (well, it would only be released after exiting the editor). If the resource is deleted, it's even worse, because the object will still have a reference from the plugin.
Why would it be a problem only when stored in an array? (I don't do that anyways).

**Comment by GlaDOSik** [comment]
But you have only one reference to edited resource, right? So if user selects the different one, the old one is lost. That means even if user deletes the resource, you have only one 1024x1024 image in memory. It's not a big deal. But if you append every selected resource to array and never call clear on it, that would be a problem. If you're still worried about that one garbage image, yeah, you can set reference to null in `make_visible(false)`, as you wrote.

**Comment by Zylann** [comment]
I'm probably going to do that, since I want to do things cleanly. It's a bit sad we would have to wait for an actual big issue to fix this while we could just make things clean in the first place (especially when it's obvious like here :p)

**Comment by Noshyaar** [comment]
First of all thank you for your report and sorry for the delay.

We released Godot 3.0 in January 2018 after 18 months of work, fixing many old issues either directly, or by obsoleting/replacing the features they were referring to.

We still have hundreds of issues whose relevance/reproducibility needs to be checked against the [current stable version](https://godotengine.org/download), and that's where you can help us.
**Could you check if the issue that you described initially is still relevant/reproducible in Godot 3.0 or any newer version, and comment about it here?**

For bug reports, please also make sure that the issue contains detailed steps to reproduce the bug and, if possible, a zipped project that can be used to reproduce it right away. This greatly speeds up debugging and bugfixing tasks for our contributors.

Our Bugsquad will review this issue more in-depth in 15 days, and potentially close it if its relevance could not be confirmed.

Thanks in advance.

*Note: This message is being copy-pasted to many "stale" issues (90+ days without activity). It might happen that it is not meaningful for this specific issue or appears oblivious of the issue's context, if so please comment to notify the Bugsquad about it.*

**Comment by Zylann** [comment]
Still no explicit way (as in "robust") to remove reference to the edited object when that object is no longer selected.

**Comment by Zylann** [comment]
A signal where?
I haven't worked much on this kind of plugins recently, so I'm re-reading the issue:
So far I'm mostly thinking that `make_visible(false)` is the only thing I can rely on to cleanup internal variables of the plugin. So there is a solution technically, but because `make_visible(false)` would be used to do that, the name went a bit misleading:
- `edit(obj)` is called when a supported object is selected,
- while `make_visible` is *assumed* to be called when none of those the plugins can handle are selected. But this is not guaranteed by the API (as it just tells the plugin to hide) and you have to make that assumption.
Mostly naming and doc stuff I'd say.

**Comment by akien-mga** [comment]
> A signal where?

I mean that a signal could be emitted when the object is deselected. Or when the selection changes, with a parameter pointing to the newly selected object (null if deselected).

**Comment by Zylann** [comment]
Just hit this again.
When the user selects an object my plugin cannot handle, I expect to be able to cleanup my plugin's UI (and everything it references).

So I started by doing that when `make_visible(false)` is called. Except this time it doesn't work:
My plugin invokes `EditorInterface::inspect_object` on a resource it can also handle. But Godot still calls `make_visible(false)` on apparently ALL plugins when I do that, only to show it back due to my plugin actually handling that resource.

So two awful things happen:
- My UI is cleaned up and rebuilt entirely to the identical (it's a `GraphEdit`), which is not only a waste, but also triggers an error, because the signal that led me to call `inspect_object` came from a `GraphNode`, and now it's getting destroyed.
- The fact the UI is rebuilt makes it loose all context, so selection is not visible anymore and box-select became super glitchy.

Now I decided to give up and not clean anything up. But even then, I'm blocked by the same problem: Godot still hides and shows my UI simultaneously due to blindly calling `make_visible(false)` and then `make_visible(true)` immediately after. That makes the `GraphEdit` glitchy when selecting nodes, it doesn't get mouse release button events and nodes stick to it...

**Comment by MJacred** [comment]
@Zylann: I think this was fixed with https://github.com/godotengine/godot/pull/71770 (Godot 4 only), which introduced an "editing owner" logic.

Can you confirm?

**Comment by Zylann** [comment]
Several things happened since:

- `edit(null)` occurs when selecting a non-handled object, which I suppose can be used to clean up the UI. Indeed looks like that's what #71770 introduced. However I havent re-tested the case of closing the scene I mentionned at the end of my last comment, as this case does not involve "editing" something else, it's just closing the open scene.
- As a result of 71770, what I was doing by handling more than one object in my plugin is actually forbidden now, errors are printed, which I had to workaround by no longer handling multiple objects. But I still have to use the inspector to inspect custom objects though, so I workaround this by actively ignoring `edit(null)` so the "awful" things I mentionned don't happen (#73650). 
- I generally became more lenient on what keeps resource references alive, even though I believe it can lead to unexpected situations/memory hogs. Maybe I'll come back to this issue if I hit this again.
